### PR TITLE
special case the buggy links in akalin's sigchain

### PR DESCRIPTION
**Pull request by oconnor663**
https://github.com/keybase/libkeybase-js/issues/5

@maxtaco

### CODE REVIEW

**Comment by oconnor663** [comment]
I've gone ahead and push a test vector here: https://github.com/keybase/keybase-test-vectors/commit/deddbe2f2bee5fc4a8d82cf40b96a4fea9fa9a73

File: src/sigchain/sigchain.iced [hunk 1]
```diff
@@ -425,8 +446,19 @@ exports.SigChain = class SigChain
 
     # This constructor checks that the link is internally consistent: its
     # signature is valid and belongs to the key it claims, and the same for any
-    # reverse sigs.
-    await ChainLink.parse {sig_blob, key_state: @_key_state, sig_cache}, esc defer link
+    # reverse sigs. Note that we need to explicitly check the error here, in
+    # case the sig is a known buggy one that we intend to skip.
+    await ChainLink.parse {sig_blob, key_state: @_key_state, sig_cache, log}, defer err, link
+    if err?
```
Comment (maxtaco) [inline src/sigchain/sigchain.iced]: Why is it necessary to short-circuit chain integrity here?  As far as I know, Fred's chain still parses fine w/r/t prev links.

File: src/sigchain/sigchain.iced [hunk 2]
```diff
@@ -76,6 +76,19 @@ bad_whitespace_sig_ids = {
   "db9a0afaab297048be0d44ffd6d89a3eb6a003256426d7fd87a60ab59880f8160f" : true
 }
 
+# Sometimes a bug in our software leads to signatures that we will never be
+# able to validate. (Sorry Fred!) We can't just delete these, because that
```
Comment (maxtaco) [inline src/sigchain/sigchain.iced]: BTW, these links weren't buggy.  We just changed the format shortly thereafter and didn't want to keep backwards-compatible code for something that never saw the light of day... Erm, nearly didn't see the light of day...

**Comment by maxtaco** [comment]
Thanks for this, but I had some general questions about approach here. Let's discuss before the merge.

File: src/sigchain/sigchain.iced [hunk 3]
```diff
@@ -200,6 +214,13 @@ class ChainLink
     await sig_eng.get_body_and_unverified_payload(
       {armored: sig_blob.sig}, esc defer sig_body, unverified_buffer)
     sig_id = kbpgp.hash.SHA256(sig_body).toString("hex") + SIG_ID_SUFFIX
+
+    # This is the earliest possible point we can identify known buggy sigs,
+    # right after we've computed the SigID. Callers should catch this error.
```
Comment (maxtaco) [inline src/sigchain/sigchain.iced]: I think it's also fine to collect payload hashes from the buggy chainlinks. This would prevent the workaround below.

**Comment by oconnor663** [comment]
Went with a different approach to avoid the problem max pointed out. Now we still parse the links, but specifically skip the reverse sig check. Then later after verifying the seqno/payload hash, we short circuit before both 1) checking the status of the signing sibkey, and 2) applying the contents of the bad link.

@maxtaco, could you please give this another look?

File: src/sigchain/sigchain.iced [hunk 4]
```diff
@@ -179,8 +189,10 @@ class ChainLink
     # Make sure the KID from the server matches the payload, and that any
     # payload PGP fingerprint also matches the KID.
     await @_check_payload_against_server_kid {sig_blob, payload, key_state}, esc defer()
-    # Check any reverse signatures.
-    await @_check_reverse_signatures {payload, key_state}, esc defer()
+    # Check any reverse signatures. For links where we skip this step, we will
+    # ignore their contents later.
+    if sig_id not of known_buggy_reverse_sigs
```
Comment (maxtaco) [inline src/sigchain/sigchain.iced]: Use: `known_buggy_reverse_sigs[sig_id]`

File: src/sigchain/sigchain.iced [hunk 5]
```diff
@@ -448,6 +460,15 @@ exports.SigChain = class SigChain
         cb new E.NotLatestSubchainError("Found a later subchain with eldest kid #{link.eldest_kid}")
         return
 
+    # Links with bad reverse sigs still have to have valid payload hashes and
+    # seqnos, but their contents are ignored, and their signing keys might not
+    # be valid sibkeys (because the delegating links of those sibkeys might
+    # also have been bad). Short-circuit here, after checking the link position
+    # but before checking the validity of the signing key.
+    if link.sig_id of known_buggy_reverse_sigs
```
Comment (maxtaco) [inline src/sigchain/sigchain.iced]: Same as above.

**Comment by maxtaco** [comment]
:+1:

**Comment by oconnor663** [comment]
stopped using `of`

**Comment by oconnor663** [comment]
shipped in node client version 0.8.22

**Pull request merged**